File: Bencode/Bencode.py
# test789a : USAG-Lib bencode

import base64
import unicodedata

splitable = ["!", "@", "#", "$", "%", "^", "&", "*", "~", "|"]

def Encode64(data: bytes, spliter: str = "", linenum: int = 40, colnum: int = 10) -> str:
    raw = "" if len(data) == 0 else base64.b64encode(data).decode('ascii')
    if spliter == "":
        return raw
    if spliter not in splitable:
        raise Exception("invalid spliter option")

    # split raw text
    lines = [ ]
    for i in range(0, len(raw), linenum):
        lines.append( raw[i:i+linenum] )
    cols = [ ]
    for i in range(0, len(lines), colnum):
        cols.append( lines[i:i+colnum] )

    # assemble text
    res = [f"{spliter}START{spliter}\n"]
    for i in range(0, len(cols)):
        res.append(f"{spliter}{i+1}/{len(cols)}{spliter}\n{'\n'.join(cols[i])}\n")
    res.append(f"{spliter}END{spliter}")
    return "".join(res)

def Decode64(data: str, spliter: str = "") -> bytes:
    data = data.replace("\t", "").replace("\r", "").replace("\n", "").replace(" ", "")
    if spliter != "" and (spliter not in splitable):
        raise Exception("invalid spliter option")

    # remove comments
    if spliter != "":
        temp, appendable, pos = [ ], True, 0
        while pos < len(data):
            if data[pos] == spliter:
                appendable = not appendable
            elif appendable:
                temp.append(data[pos])
            pos = pos + 1
        data = "".join(temp)

    if data == "":
        return b""
    return base64.b64decode(data)

def NormPW(pw: str) -> bytes:
    if pw == "":
        return b""
    return unicodedata.normalize('NFC', pw).encode('utf-8')

File: Bencode/Bencode.js
// test789b : USAG-Lib bencode
const isNode = (typeof window === 'undefined');

// ===== helpers =====
const splitable = new Set(["!", "@", "#", "$", "%", "^", "&", "*", "~", "|"]);

function _toBase64(uint8Array) {
    if (isNode) {
        return Buffer.from(uint8Array).toString('base64');
    } else {
        let binary = '';
        const len = uint8Array.byteLength;
        for (let i = 0; i < len; i++) binary += String.fromCharCode(uint8Array[i]);
        return btoa(binary);
    }
}

function _fromBase64(str) {
    if (isNode) {
        return new Uint8Array(Buffer.from(str, 'base64'));
    } else {
        const binary = atob(str);
        const len = binary.length;
        const bytes = new Uint8Array(len);
        for (let i = 0; i < len; i++) bytes[i] = binary.charCodeAt(i);
        return bytes;
    }
}

// ===== main =====
export function Encode64(data, spliter = "", linenum = 40, colnum = 10) {
    // convert data to Uint8Array
    if (typeof data === 'string') {
        data = new TextEncoder().encode(data);
    } else if (!(data instanceof Uint8Array)) {
        data = new Uint8Array(data || []);
    }

    // encode to base64
    let raw = "";
    if (data.length > 0) {
        raw = _toBase64(data);
    }

    // check spliter
    if (spliter === "") {
        return raw;
    }
    if (!splitable.has(spliter)) {
        throw new Error("invalid spliter option");
    }

    // split text
    const lines = [];
    for (let i = 0; i < raw.length; i += linenum) {
        lines.push(raw.slice(i, i + linenum));
    }
    const cols = [];
    for (let i = 0; i < lines.length; i += colnum) {
        cols.push(lines.slice(i, i + colnum));
    }

    // assemble text
    let res = `${spliter}START${spliter}\n`;
    const totalCols = cols.length;
    for (let i = 0; i < totalCols; i++) {
        res += `${spliter}${i + 1}/${totalCols}${spliter}\n${cols[i].join('\n')}\n`;
    }
    res += `${spliter}END${spliter}`;
    return res;
}

export function Decode64(data, spliter = "") {
    data = data.replace(/[\r\n \t]/g, "");
    if (spliter !== "" && !splitable.has(spliter)) {
        throw new Error("invalid spliter option");
    }

    // remove comment
    if (spliter !== "") {
        const parts = data.split(spliter);
        let pureData = "";
        for (let i = 0; i < parts.length; i += 2) {
            pureData += parts[i];
        }
        data = pureData;
    }

    // decode
    if (data === "") {
        return new Uint8Array(0);
    }
    return _fromBase64(data);
}

/**
 * @param {string} pw string
 * @returns {Uint8Array} NFC UTF-8 bytes
 */
export function NormPW(pw) {
    if (!pw) return new Uint8Array(0);
    const encoder = new TextEncoder();
    return encoder.encode(pw.normalize('NFC')); 
}